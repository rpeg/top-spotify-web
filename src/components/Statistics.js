import React from 'react';
import { useSelector } from 'react-redux';
import { mean, mode } from 'mathjs';

import { Row, Col } from 'react-bootstrap';
import Radar from './Radar';
import Statistic from './Statistic';
import RadialBar from './RadialBar';
import Donut from './Donut';

const getMode = (items) => mode(items)[0];
const getMean = (items) => mean(items);

const getKey = (pitch) => {
  const letters = [
    'C',
    'C#',
    'D',
    'D#',
    'E',
    'F',
    'F#',
    'G',
    'G#',
    'A',
    'A#',
    'B',
  ];

  return letters[pitch % 12];
};

const Statistics = ({ features, tracks }) => {
  const statsOptions = useSelector((state) => state.statsOptions);

  return (
    <div>
      <Row>
        {statsOptions.includes('key') && (
        <Col>
          <Statistic
            title="Most Frequent Key"
          >
            <h4>{getKey(getMode(features.map((f) => f.key)))}</h4>
          </Statistic>
        </Col>
        )}
        {statsOptions.includes('bpm') && (
        <Col>
          <Statistic
            title="Average BPM"
          >
            <h4>{getMean(features.map((f) => f.tempo))}</h4>
          </Statistic>
        </Col>
        )}
      </Row>
      <Row>
        {statsOptions.includes('scale') && (
        <Col>
          <Statistic
            title="Scale"
          >
            <RadialBar items={features.map((f) => f.mode)} />

          </Statistic>
        </Col>
        )}
        {statsOptions.includes('decades') && (
        <Col>
          <Statistic
            title="Decades"
          >
            <Donut
              items={tracks.map((t) => `${t.album.release_date[2]}0s`)}
            />
          </Statistic>
        </Col>
        )}
      </Row>
      <Row>
        {statsOptions.includes('features') && (
        <Col xs={12}>
          <Radar features={features} />
        </Col>
        )}
      </Row>
    </div>
  );
};

export default Statistics;
