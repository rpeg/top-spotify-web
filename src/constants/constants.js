export const SET_USER = 'SET_USER';
export const SET_TIME_RANGE = 'SET_TIME_RANGE';
export const SET_ARTIST_COUNT = 'SET_ARTIST_COUNT';
export const SET_TRACK_COUNT = 'SET_TRACK_COUNT';
export const SET_GENRE_COUNT = 'SET_GENRE_COUNT';
export const REQUEST_ARTISTS = 'REQUEST_ARTISTS';
export const REQUEST_TRACKS = 'REQUEST_TRACKS';
export const REQUEST_FEATURES = 'REQUEST_FEATURES';
export const RECEIVE_ARTISTS = 'RECEIVE_ARTIST';
export const RECEIVE_TRACKS = 'RECEIVE_TRACKS';
export const RECEIVE_FEATURES = 'RECEIVE_FEATURES';

// TODO increase N once Spotify allows for more top artists/tracks
export const ARTIST_REQ_LIMIT = 50;
export const TRACK_REQ_LIMIT = 50;
export const N_ARTISTS = ARTIST_REQ_LIMIT;
export const N_TRACKS = TRACK_REQ_LIMIT;
export const FEATURE_REQ_LIMIT = 100;

export const TimeRanges = {
  LONG: { range: 'long_term', title: 'All-time' },
  MEDIUM: { range: 'medium_term', title: 'Last Six Months' },
  SHORT: { range: 'short_term', title: 'Recent' },
};
