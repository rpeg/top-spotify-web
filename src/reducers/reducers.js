import { combineReducers } from 'redux';
import {
  SET_USER,
  SET_TIME_RANGE,
  TimeRanges,
  REQUEST_ARTISTS,
  RECEIVE_ARTISTS,
  REQUEST_TRACKS,
  RECEIVE_TRACKS,
  REQUEST_FEATURES,
  RECEIVE_FEATURES,
  SET_ARTIST_COUNT,
  SET_TRACK_COUNT,
  SET_GENRE_COUNT,
} from '../constants/constants';

const { LONG } = TimeRanges;

function user(state = {}, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, user: action.user };
    default:
      return state;
  }
}

function timeRange(state = LONG, action) {
  switch (action.type) {
    case SET_TIME_RANGE:
      return { ...state, timeRange: action.timeRange };
    default:
      return state;
  }
}

function artistCount(state = 10, action) {
  switch (action.type) {
    case SET_ARTIST_COUNT:
      return { ...state, artistCount: action.artistCount };
    default:
      return state;
  }
}

function trackCount(state = 15, action) {
  switch (action.type) {
    case SET_TRACK_COUNT:
      return { ...state, trackCount: action.trackCount };
    default:
      return state;
  }
}

function genreCount(state = 40, action) {
  switch (action.type) {
    case SET_GENRE_COUNT:
      return { ...state, genreCount: action.genreCount };
    default:
      return state;
  }
}

function artists(state = {}, action) {
  switch (action.type) {
    case REQUEST_ARTISTS:
      return { ...state, isFetching: true };
    case RECEIVE_ARTISTS:
      return { ...state, isFetching: false, items: action.items };
    default:
      return state;
  }
}

function artistsByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_ARTISTS:
    case REQUEST_ARTISTS:
      return { ...state, [action.timeRange]: artists(state[action.timeRange], action) };
    default:
      return state;
  }
}

function tracks(state = {}, action) {
  switch (action.type) {
    case REQUEST_TRACKS:
      return { ...state, isFetching: true };
    case RECEIVE_TRACKS:
      return { ...state, isFetching: false, items: action.tracks };
    default:
      return state;
  }
}

function tracksByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_TRACKS:
    case REQUEST_TRACKS:
      return { ...state, [action.timeRange]: tracks(state[action.timeRange], action) };
    default:
      return state;
  }
}

function features(state = {}, action) {
  switch (action.type) {
    case REQUEST_FEATURES:
      return { ...state, isFetching: true };
    case RECEIVE_FEATURES:
      return { ...state, isFetching: false, items: action.features };
    default:
      return state;
  }
}

const rootReducer = combineReducers({
  user,
  timeRange,
  artistCount,
  trackCount,
  genreCount,
  artists,
  artistsByTimeRange,
  tracks,
  tracksByTimeRange,
  features,
});

export default rootReducer;
