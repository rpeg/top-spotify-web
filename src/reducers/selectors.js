export const selectHasClickedCreate = (state) => state.hasClickedCreate;
export const selectUser = (state) => state.user;
export const selectTimeRangeName = (state) => state.timeRangeName;
export const selectGenreCount = (state) => state.genreCount;
export const selectArtistCount = (state) => state.artistCount;
export const selectTrackCount = (state) => state.trackCount;
export const selectStatsOptions = (state) => state.statsOptions;
export const selectDisplayProfile = (state) => state.displayProfile;
export const selectOptimizeTracks = (state) => state.optimizeTracks;
export const selectArtists = (state) => state.artists;
export const selectArtistsByCurrentTimeRange = (state) => state.artistsByTimeRangeName[state.timeRangeName];
export const selectTracks = (state) => state.tracks;
export const selectTracksByCurrentTimeRange = (state) => state.tracksByTimeRangeName[state.timeRangeName];
export const selectFeatures = (state) => state.features;
export const selectFeaturesByCurrentTimeRange = (state) => state.featuresByTimeRangeName[state.timeRangeName];
