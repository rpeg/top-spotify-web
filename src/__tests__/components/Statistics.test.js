import React from 'react';
import { shallow } from 'enzyme';

import Statistics from '../../components/Statistics';
import * as selectors from '../../reducers/selectors';
import { DEFAULT_STATS_OPTIONS } from '../../constants/constants';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((fn) => fn()),
  useDispatch: () => jest.fn(),
}));

const setup = (statsOptions = DEFAULT_STATS_OPTIONS) => {
  jest.spyOn(selectors, 'selectStatsOptions').mockReturnValue(statsOptions);
};

describe('Statistics', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should not log errors in console', () => {
    setup();
    const wrapper = shallow(<Statistics />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    setup();
    const wrapper = shallow(<Statistics />);
    expect(wrapper).toMatchSnapshot();
  });
});
