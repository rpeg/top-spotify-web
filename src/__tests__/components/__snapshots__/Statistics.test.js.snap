// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics renders correctly 1`] = `
<div>
  <ForwardRef
    noGutters={false}
  />
  <ForwardRef
    noGutters={false}
  >
    <Col
      md={4}
      xs={12}
    >
      <Statistic
        title="Decades"
      >
        <ForwardRef
          className="justify-content-center"
          noGutters={false}
        >
          <Col
            md={12}
            xs={9}
          >
            <Donut
              items={Array []}
            />
          </Col>
        </ForwardRef>
      </Statistic>
    </Col>
    <Col
      md={8}
      xs={12}
    >
      <Statistic
        title="Features"
      >
        <Radar
          features={Array []}
        />
      </Statistic>
    </Col>
  </ForwardRef>
</div>
`;
